### PR TITLE
Use the determinant formula for calculating the orientation of a polygon

### PR DESCRIPTION
**Pull request by DaveCTurner**
The method for working out whether a polygon is clockwise or anticlockwise is mostly correct but doesn't work in some rare cases such as the included test case. This PR changes it to determine the orientation from the direction of the cross product of two of the edges of the polygon at a point on its convex hull, which should work more reliably.

### CODE REVIEW

**Comment by DaveCTurner** [comment]
@nknize could you have a look at this please?

**Review by nknize — COMMENTED** [review]
Sorry for the delay. I'm good with this. As a side note I can make the `orient` method (https://github.com/apache/lucene-solr/blob/df9b88443430a85c4bb00af63ff1da911584e6f5/lucene/core/src/java/org/apache/lucene/geo/Polygon2D.java#L466) public in the Lucene GeoAPI. Might make a note to replace this method once that's done so we don't forget.

**Comment by DaveCTurner** [comment]
Ok, I added a TODO note as requested. You said "I'm good with this" but didn't give an "approved" review - could you?

**Comment by DaveCTurner** [comment]
Pinging @elastic/es-search-aggs (and @nknize 🙂)

**Comment by DaveCTurner** [comment]
I think this is good to merge but needs a final LGTM - could someone from the Geo team take a look?

**Comment by DaveCTurner** [comment]
Could someone from @elastic/es-search-aggs take a look at this please?

**Comment by DaveCTurner** [comment]
@nknize @elastic/es-search-aggs could you take a last look at this please?

File: server/src/main/java/org/elasticsearch/common/geo/builders/PolygonBuilder.java
Comment (jtibshirani) [inline server/src/main/java/org/elasticsearch/common/geo/builders/PolygonBuilder.java]: I have a couple questions about this assertion:
- I think the determinant could be 0 if the three points lie in a straight line. Is my understanding correct that this won't happen because of the way we choose `top`? It might be nice to add a quick note to explain.
- Are we concerned about numeric underflow here? I'm curious as to how we handle numeric issues in other parts of the geo code (in particular, whether we should throw a runtime exception when an issue is detected).

**Review by jtibshirani — COMMENTED** [review]
Hi @DaveCTurner, I'm sorry you've been waiting so long for a review. I’m new to the geo code, but did my best to take a look and left a question.

It looks like in the latest version of lucene, the `orient` method is now public: https://github.com/apache/lucene-solr/blob/master/lucene/core/src/java/org/apache/lucene/geo/GeoUtils.java#L185. We're upgrading to the latest lucene snapshot now (#32390), so it may be nice to cut over to that when the upgrade goes in. We could also do it in a follow-up if you (understandably) don't want to wait.

**Comment by DaveCTurner** [comment]
Thanks for spotting that this has landed in Lucene @jtibshirani. #32390 is merged so I've updated this PR to defer to that.

> I have a couple questions about this assertion:
> - I think the determinant could be 0 if the three points lie in a straight line. Is my understanding correct that this won't happen because of the way we choose top? It might be nice to add a quick note to explain.

On closer inspection the points could in fact be collinear at this point, but only if `top` is the last of them, in which case the edges intersect so the shape gets thrown out later on in its construction. I think it's better to reject that here as the message can be made clearer, so I've done that.

> - Are we concerned about numeric underflow here? I'm curious as to how we handle numeric issues in other parts of the geo code (in particular, whether we should throw a runtime exception when an issue is detected).

It seems not - the Lucene implementation is the same as the one written here, although at least it has a link to a more robust implementation:

```
  // see the "Orient2D" method described here:
  // http://www.cs.berkeley.edu/~jrs/meshpapers/robnotes.pdf
  // https://www.cs.cmu.edu/~quake/robust.html
  // Note that this one does not yet have the floating point tricks to be exact!
```

(the linked `Orient2D` method is also not 100% exact, I think, because it assumes support for unbounded exponents, but it's the best one I have come across). I also ran into a similar issue in JTS when investigating your collinearity question. I guess the argument is that for geospatial data this kind of error happens at the scale of µm, and some continents are moving further than that every hour, so it doesn't make sense to work with shapes that are so close to degeneracy anyway. Not a very satisfying answer.

File: server/src/main/java/org/elasticsearch/common/geo/builders/PolygonBuilder.java
Comment (jtibshirani) [inline server/src/main/java/org/elasticsearch/common/geo/builders/PolygonBuilder.java]: I think there's an extra 'if so' here.

**Review by jtibshirani — APPROVED** [review]
Thanks for looking into the collinearity + underflow issues! It looks good to me.

**Pull request merged**
